src/lib.rs

```diff
@@ -1,14 +1,180 @@
+    tree_sitter_grep::run_with_callback(
+        tree_sitter_grep_args,
+        |capture_info, file_contents, path| {
+            let (rule_index, rule_listener_index) =
+                aggregated_queries.pattern_index_lookup[capture_info.pattern_index];
+            let rule = &resolved_rules[rule_index];
+            let listener = &rule.listeners[rule_listener_index];
+            (listener.on_query_match)(
+                &capture_info.node,
+                &QueryMatchContext::new(path, file_contents, rule, &reported_any_violations),
+            );
+        },
+    )
+    .unwrap();
+    if reported_any_violations.load(Ordering::Relaxed) {
+        process::exit(1);
+    } else {
+        process::exit(0);
+    }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+type RuleIndex = usize;
+type RuleListenerIndex = usize;
 
-    #[test]
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
+struct AggregatedQueries {
+    pattern_index_lookup: Vec<(RuleIndex, RuleListenerIndex)>,
+    #[allow(dead_code)]
+    query: Query,
+    query_text: String,
+}
+
+impl AggregatedQueries {
+    pub fn new(resolved_rules: &[ResolvedRule], context: &Context) -> Self {
+        let mut pattern_index_lookup: Vec<(RuleIndex, RuleListenerIndex)> = Default::default();
+        let mut aggregated_query_text = String::new();
+        for (rule_index, resolved_rule) in resolved_rules.into_iter().enumerate() {
+            for (rule_listener_index, rule_listener) in resolved_rule.listeners.iter().enumerate() {
+                for _ in 0..rule_listener.query.pattern_count() {
+                    pattern_index_lookup.push((rule_index, rule_listener_index));
+                }
+                let use_capture_name =
+                    &rule_listener.query.capture_names()[rule_listener.capture_index as usize];
+                let query_text_with_unified_capture_name =
+                    regex!(&format!(r#"@{use_capture_name}\b"#)).replace_all(
+                        &rule_listener.query_text,
+                        CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT,
+                    );
+                assert!(
+                    matches!(query_text_with_unified_capture_name, Cow::Owned(_),),
+                    "Didn't find any instances of the capture name to replace"
+                );
+                aggregated_query_text.push_str(&query_text_with_unified_capture_name);
+                aggregated_query_text.push_str("\n\n");
+            }
+        }
+        let query = Query::new(context.language, &aggregated_query_text).unwrap();
+        assert!(query.pattern_count() == pattern_index_lookup.len());
+        Self {
+            pattern_index_lookup,
+            query,
+            query_text: aggregated_query_text,
+        }
     }
 }
+
+fn get_rules() -> Vec<Rule> {
+    vec![no_default_default_rule(), no_lazy_static_rule()]
+}
+
+fn no_default_default_rule() -> Rule {
+    RuleBuilder::default()
```

Comment:
The builder pattern seemed appropriate for these but I could also see adding an (optional) layer on top of that for people who wanted a slightly "slimmer" syntax eg:
```
rule! {
    name => no_default_default,
    create => |context| {
        vec![
            rule_listener! {
                query => r#"..."#,
                capture_name => "c",
                on_query_match => |node, query_match_context| {
                    ...
                }
            },
            ...
        ]
    }
}
```

Reply:
This theme keeps coming up of tooling "pulling up short" of really trying to parse the contents of macro invocations and consumers suffering for it (the thing I had to do for `tree-sitter-grep` of forking the Rust tree-sitter grammar to try and parse "typical" macro-invocation arguments; a similar thing came up here with doing `syn`-based traversal of parsed AST's where it "wasn't working as expected" without some extra "convincing" to traverse into macro contents; and like you're saying `rustfmt` not formatting macro invocation contents that it kind of seems like it should be able to format)

Ya I'm sort of into the more "magical" macro syntax for this use case of trying to make rule authorship feel "smooth" (and ESLint-like) but I'm assuming it won't be everyone's style

In which case (in terms of how the rule-authoring API's are currently organized) (a) you can always write the more boilerplate-y but "just Rust" version and/or (b) there's really no reason there couldn't be _alternate_ macro syntaxes (whether "official" or not) (since it's just a way of generating "publicly-writable Rust code") that are maybe a bit less "magical"?